rfcs/convergence/make-overrides.md

```diff
@@ -0,0 +1,184 @@
+
+Proposed API solve the problem in a "vandal" way 🪓 Matchers are moved to user's scope thus we can have all definitions in a single `makeStyles()` call => we have a single React hook. See a modified snippet below:
+
+```tsx
+import { ax, makeOverrides } from "@fluentui/react-make-styles";
+
+const useStyles = makeStyles({
+  /* 👍 no matchers, no need to execute on each render */
+  root: { color: "red" },
+  rootPrimary: { color: "blue" },
+
+  /* 👍 styles for each slot are defined together (not a requirement) */
+  icon: { background: "black" },
+  iconPrimary: { background: "white" }
+});
+
+function Component() {
+  /* 👍 a single call of React hook */
+  const classes = useStyles();
+
+  return (
+    <div
+      className={ax(
+        classes.root /* The concept of matching is replaced with selective classname concat */,
+        props.primary && classes.rootPrimary,
+        props.className
+      )}
+    >
+      {props.icon && (
+        <div
+          className={ax(
+            classes.icon,
+            props.primary && classes.iconPrimary,
+            props.icon.className
+          )}
+        />
+      )}
+    </div>
+  );
+}
+```
+
+### Pros and Cons
+
+These changes result in performance improvements in `makeOveriddes`:
+
+- we are not executing matchers on each render inside styling functions
+- we are not executing or merging styles until slots are present
+- less React hooks to call for slots
+
+Surprisingly, this also moves us closer to [CSS Modules](https://github.com/css-modules/css-modules). The same snippet with CSS Modules, for example:
+
+```tsx
+import cx from "classnames";
+import * as classes from "./Component.css"; // the same styles can be written in CSS
+
+function Component() {
+  return (
+    <div
+      className={cx(
+        classes.root,
+        props.primary && classes.rootPrimary,
+        props.className
+      )}
+    >
+      {props.icon && (
+        <div
+          className={cx(
+            classes.icon,
+            props.primary && classes.iconPrimary,
+            props.icon.className
+          )}
+        />
+      )}
+    </div>
+  );
+}
+```
+
+In [microsoft/fluentui#16923](https://github.com/microsoft/fluentui/pull/16923)  `Avatar`, from `@fluentui/react-avatar` (check `packages/react-avatar/src/components/Avatar/useAvatarStyles.ts`), was used to validate performance improvements and explore potential issues. It is uncertain if/how these issues are critical, but there are two things that should be mentioned:
+
+```tsx
+export const useAvatarStyles = (state: AvatarState): AvatarState => {
+  const classes = useStyles();
+
+  state.className = ax(
+    classes.root,
+
+    //
+
+    state.size === 20 && classes.rootShape20,
+    state.size === 24 && classes.rootShape24,
+    /* ...many selectors... */
+    state.size === 128 && classes.rootShape128
+  );
+
+  return state;
+};
+```
```

Comment:
```suggestion
```tsx
const useStyles = makeOverrides({
  /* ... many definitions ... */

  rootShape20: { width: '20px', height: '20px' },
  rootShape24: { width: '24px', height: '24px' },
  rootShape28: { width: '28px', height: '28px' },
})

export const useAvatarStyles = (state: AvatarState): AvatarState => {
  const classes = useStyles();

  state.className = ax(
    classes.root,

    // 👎 Matchers have been moved to ax() calls, it looks a bit verbose
    //    (in previous implementation matchers have been close to styles)
    // 👎 It might be tricky find proper names to express definition names
    //    (we can end with "rootPrimaryCirclularGhostEtc.")

    state.size === 20 && classes.rootShape20,
    state.size === 24 && classes.rootShape24,
    /* ... many selectors ... */
    state.size === 128 && classes.rootShape128
  );

  return state;
};
```
```